superbench/benchmarks/micro_benchmarks/directx_third_party/DeviceResources.cpp

```diff
@@ -0,0 +1,670 @@
+//
+// DeviceResources.cpp - A wrapper for the Direct3D 12 device and swapchain
+//
+
+#include "DeviceResources.h"
+
+using namespace DirectX;
+using namespace DX;
+
+using Microsoft::WRL::ComPtr;
+
+#ifdef __clang__
+#pragma clang diagnostic ignored "-Wcovered-switch-default"
+#pragma clang diagnostic ignored "-Wswitch-enum"
+#endif
+
+#pragma warning(disable : 4061)
+
+namespace {
+inline DXGI_FORMAT NoSRGB(DXGI_FORMAT fmt) noexcept {
+    switch (fmt) {
+    case DXGI_FORMAT_R8G8B8A8_UNORM_SRGB:
+        return DXGI_FORMAT_R8G8B8A8_UNORM;
+    case DXGI_FORMAT_B8G8R8A8_UNORM_SRGB:
+        return DXGI_FORMAT_B8G8R8A8_UNORM;
+    case DXGI_FORMAT_B8G8R8X8_UNORM_SRGB:
+        return DXGI_FORMAT_B8G8R8X8_UNORM;
+    default:
+        return fmt;
+    }
+}
+
+inline long ComputeIntersectionArea(long ax1, long ay1, long ax2, long ay2, long bx1, long by1, long bx2,
+                                    long by2) noexcept {
+    return std::max(0l, std::min(ax2, bx2) - std::max(ax1, bx1)) *
+           std::max(0l, std::min(ay2, by2) - std::max(ay1, by1));
+}
+} // namespace
+
+// Constructor for DeviceResources.
+DeviceResources::DeviceResources(DXGI_FORMAT backBufferFormat, DXGI_FORMAT depthBufferFormat, UINT backBufferCount,
+                                 D3D_FEATURE_LEVEL minFeatureLevel, unsigned int flags) noexcept(false)
+    : m_backBufferIndex(0), m_fenceValues{}, m_rtvDescriptorSize(0), m_screenViewport{}, m_scissorRect{},
+      m_backBufferFormat(backBufferFormat), m_depthBufferFormat(depthBufferFormat), m_backBufferCount(backBufferCount),
+      m_d3dMinFeatureLevel(minFeatureLevel), m_window(nullptr), m_d3dFeatureLevel(D3D_FEATURE_LEVEL_11_0),
+      m_dxgiFactoryFlags(0), m_outputSize{0, 0, 1, 1}, m_colorSpace(DXGI_COLOR_SPACE_RGB_FULL_G22_NONE_P709),
+      m_options(flags), m_deviceNotify(nullptr) {
+    if (backBufferCount < 2 || backBufferCount > MAX_BACK_BUFFER_COUNT) {
+        throw std::out_of_range("invalid backBufferCount");
+    }
+
+    if (minFeatureLevel < D3D_FEATURE_LEVEL_11_0) {
+        throw std::out_of_range("minFeatureLevel too low");
+    }
```

Comment:
do you think it is good to throw these exceptions in the constructor? Do we handle them in the code where we create the new instances of this class?

Reply:
currently I will handle all exceptions in the main loop, do you think it's ok?

Reply:
well, I still don't like the design like this, but it is OK if you have handled it. :smile: